File: frontend/src/api/imports.ts
import { apiRequest } from "./client";
import type { UUID } from "../types/api";

/** 后端 ImportMappingRequest：四个可映射字段，blank 会被规范化为 None。 */
export interface ImportMapping {
  source_row_number?: string | null;
  external_work_order_number?: string | null;
  title?: string | null;
  content?: string | null;
}

export interface ImportPreviewResponse {
  columns: string[];
  total_rows: number;
  suggested_mapping: Record<string, string>;
}

export interface ImportResponse {
  batch_id: UUID;
  status: string;
  total_rows: number;
  successful_rows: number;
  failed_rows: number;
  duplicate_rows: number;
  checkpoint_row: number;
  idempotent: boolean;
}

/** POST /imports/preview (multipart: file + sheet_name?) */
export function previewImport(params: {
  file: File;
  sheetName?: string;
  signal?: AbortSignal;
}): Promise<ImportPreviewResponse> {
  const form = new FormData();
  form.append("file", params.file);
  if (params.sheetName) form.append("sheet_name", params.sheetName);
  return apiRequest<ImportPreviewResponse>("/imports/preview", {
    method: "POST",
    body: form,
    signal: params.signal,
  });
}

/** POST /imports (multipart: file + mapping JSON + sheet_name?) */
export function executeImport(params: {
  file: File;
  mapping: ImportMapping;
  sheetName?: string;
  signal?: AbortSignal;
}): Promise<ImportResponse> {
  const form = new FormData();
  form.append("file", params.file);
  form.append("mapping", JSON.stringify(params.mapping));
  if (params.sheetName) form.append("sheet_name", params.sheetName);
  return apiRequest<ImportResponse>("/imports", {
    method: "POST",
    body: form,
    signal: params.signal,
  });
}

File: backend/app/schemas/imports.py
from uuid import UUID

from pydantic import BaseModel, field_validator


class ImportMappingRequest(BaseModel):
    source_row_number: str | None = None
    external_work_order_number: str | None = None
    title: str | None = None
    content: str | None = None
    reported_at: str | None = None

    @field_validator(
        "source_row_number", "external_work_order_number", "title", "content", "reported_at"
    )
    @classmethod
    def blank_to_none(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None

    def as_mapping(self) -> dict[str, str]:
        return self.model_dump(exclude_none=True)


class ImportPreviewResponse(BaseModel):
    columns: tuple[str, ...]
    total_rows: int
    suggested_mapping: dict[str, str]


class ImportResponse(BaseModel):
    batch_id: UUID
    status: str
    total_rows: int
    successful_rows: int
    failed_rows: int
    duplicate_rows: int
    checkpoint_row: int
    idempotent: bool
